src/Functions/FunctionAutoregress.cpp

```diff
@@ -0,0 +1,206 @@
+#include <Columns/ColumnFunction.h>
+#include <Common/assert_cast.h>
+#include <Core/ColumnWithTypeAndName.h>
+#include <Core/TypeId.h>
+#include <DataTypes/DataTypeFunction.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <Interpreters/Context_fwd.h>
+
+#include <memory>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+inline constexpr auto function_name = "autoregress";
+
+void checkArguments(const DataTypes & arguments)
+{
+    if (arguments[0]->getTypeId() != TypeIndex::Function)
+    {
+        throw Exception(
+            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+            "The argument 1 of the function '{}' should be lambda expression",
+            function_name);
+    }
+    const DataTypeFunction * lambda = checkAndGetDataType<DataTypeFunction>(arguments[0].get());
+    const DataTypes & lambda_args = lambda->getArgumentTypes();
+    if (lambda_args.size() != 1)
+    {
+        throw Exception(
+            ErrorCodes::BAD_ARGUMENTS,
+            "The argument 1, lambda expression of the function '{}' should accept only one argument of type Float64",
+            function_name);
+    }
+    if (arguments[1]->getTypeId() != TypeIndex::UInt8)
+    {
+        throw Exception(
+            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+            "The argument 2 of the function '{}' should be UInt8",
+            function_name);
+    }
+    if (arguments[2]->getTypeId() != TypeIndex::Int64 && arguments[2]->getTypeId() != TypeIndex::Float64)
+    {
+        throw Exception(
+            ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+            "The argument 3 of the function '{}' should be Int64 or Float64",
+            function_name);
+    }
+}
+
+class ExecutableFunctionAutoregress final : public IExecutableFunction
+{
+public:
+    String getName() const override { return function_name; }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+    {
+        auto res = result_type->createColumn();
+        if (input_rows_count == 0)
+        {
+            return res;
+        }
+
+        size_t backward_step = arguments[1].column->getUInt(0);
+
+        const ColumnFunction * column_function = assert_cast<const ColumnFunction *>(arguments[0].column.get());
```

Comment:
Maybe use `const auto *`  would be better?

Reply:
I prefer to explicitly specifying class name. `auto` impacts reading code.